File: android-app/app/src/main/java/com/example/myandroidapp/DriverActivity.java
package com.example.myandroidapp;

import android.Manifest;
import android.content.pm.PackageManager;
import android.location.Location;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;

public class DriverActivity extends AppCompatActivity {

    private TextView statusText, tvGpsStatus;
    private Button goOnlineBtn;
    private LinearLayout bookingsContainer;
    private FusedLocationProviderClient fusedLocationClient;
    private LocationCallback locationCallback;
    private boolean isOnline = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_driver);

        // Bind views
        statusText = findViewById(R.id.statusText);
        tvGpsStatus = findViewById(R.id.tvGpsStatus);
        goOnlineBtn = findViewById(R.id.goOnlineBtn);
        bookingsContainer = findViewById(R.id.bookingsContainer);

        // Setup location client
        fusedLocationClient = LocationServices.getFusedLocationProviderClient(this);

        // Go Online button
        goOnlineBtn.setOnClickListener(v -> {
            if (!isOnline) {
                startLocationUpdates();
                isOnline = true;
                goOnlineBtn.setText("Go Offline");
                statusText.setText("🟢 ONLINE");
            } else {
                stopLocationUpdates();
                isOnline = false;
                goOnlineBtn.setText("Go Online");
                statusText.setText("🔘 OFFLINE");
                tvGpsStatus.setText("Waiting for location...");
            }
        });
    }

    private void startLocationUpdates() {
        if (ActivityCompat.checkSelfPermission(this,
                Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this,
                    new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, 1001);
            return;
        }

        LocationRequest locationRequest = LocationRequest.create()
                .setInterval(5000)
                .setFastestInterval(2000)
                .setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);

        locationCallback = new LocationCallback() {
            @Override
            public void onLocationResult(LocationResult locationResult) {
                if (locationResult == null) return;
                for (Location location : locationResult.getLocations()) {
                    double lat = location.getLatitude();
                    double lng = location.getLongitude();
                    tvGpsStatus.setText("📍 " + lat + ", " + lng);
                }
            }
        };

        fusedLocationClient.requestLocationUpdates(locationRequest, locationCallback, null);
    }

    private void stopLocationUpdates() {
        if (locationCallback != null) {
            fusedLocationClient.removeLocationUpdates(locationCallback);
        }
    }

    // Add a booking card to the list
    public void addBookingCard(String passengerName, String pickup, String dropoff, String status) {
        LayoutInflater inflater = LayoutInflater.from(this);
        View card = inflater.inflate(R.layout.item_booking, bookingsContainer, false);

        TextView tvName = card.findViewById(R.id.tvPassengerName);
        TextView tvPickup = card.findViewById(R.id.tvPickupLocation);
        TextView tvDropoff = card.findViewById(R.id.tvDropoffLocation);
        TextView tvStatus = card.findViewById(R.id.tvStatus);

        tvName.setText(passengerName);
        tvPickup.setText("Pickup: " + pickup);
        tvDropoff.setText("Dropoff: " + dropoff);
        tvStatus.setText(status);

        bookingsContainer.addView(card);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        stopLocationUpdates();
    }
}

File: android-app/app/src/main/java/com/commute/ui/activities/DriverActivity.kt
package com.commute.ui.activities

import android.Manifest
import android.content.res.ColorStateList
import android.content.pm.PackageManager
import android.graphics.Color
import android.os.Bundle
import android.os.Looper
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.lifecycle.lifecycleScope
import com.commute.R
import com.commute.data.models.Booking
import com.commute.data.models.StatusUpdateRequest
import com.commute.network.ApiClient
import com.commute.network.SocketManager
import com.google.android.gms.location.*
import kotlinx.coroutines.launch

class DriverActivity : AppCompatActivity() {
    private lateinit var fusedLocationClient: FusedLocationProviderClient
    private lateinit var locationCallback: LocationCallback
    private var isOnline = false
    private var vehicleId = 1  // TODO: get from SessionManager.getVehicleId()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_driver)
        fusedLocationClient = LocationServices.getFusedLocationProviderClient(this)

        fetchBookings()  // load list immediately on screen open

        SocketManager.onNewBooking { bookingJson ->
            runOnUiThread {
                val pickup = bookingJson.optString("pickup_address", "Unknown")
                Toast.makeText(this, "New booking! Pickup: $pickup", Toast.LENGTH_LONG).show()
                fetchBookings()  // refresh list
            }
        }

        val goOnlineBtn = findViewById<Button>(R.id.goOnlineBtn)
        val statusText  = findViewById<TextView>(R.id.statusText)
        goOnlineBtn.setOnClickListener {
            if (!isOnline) {
                startLocationUpdates(); isOnline = true
                goOnlineBtn.text = "Go Offline"
                goOnlineBtn.backgroundTintList = ColorStateList.valueOf(Color.parseColor("#EF4444"))
                statusText.text = "ONLINE"; statusText.setTextColor(Color.parseColor("#10B981"))
            } else {
                stopLocationUpdates(); SocketManager.goOffline(vehicleId); isOnline = false
                goOnlineBtn.text = "Go Online"
                goOnlineBtn.backgroundTintList = ColorStateList.valueOf(Color.parseColor("#0EA5E9"))
                statusText.text = "OFFLINE"; statusText.setTextColor(Color.parseColor("#6B7FA8"))
            }
        }
    }

    private fun fetchBookings() {
        lifecycleScope.launch {
            try {
                val response = ApiClient.service.getBookings()
                if (response.isSuccessful)
                    runOnUiThread { updateBookingsList(response.body() ?: emptyList()) }
            } catch (e: Exception) { /* silent */ }
        }
    }

    private fun updateBookingsList(bookings: List<Booking>) {
        val container = findViewById<LinearLayout>(R.id.bookingsContainer)
        container.removeAllViews()
        if (bookings.isEmpty()) {
            val tv = TextView(this)
            tv.text = "No bookings yet"
            tv.setTextColor(Color.parseColor("#6B7FA8"))
            container.addView(tv); return
        }
        bookings.take(10).forEach { booking ->
            val item = layoutInflater.inflate(R.layout.item_booking, container, false)
            item.findViewById<TextView>(R.id.bRoute).text   = booking.route_name ?: "Route"
            item.findViewById<TextView>(R.id.bPickup).text  = "Pickup: ${booking.pickup_address}"
            item.findViewById<TextView>(R.id.bDropoff).text = "Drop-off: ${booking.dropoff_address}"
            item.findViewById<TextView>(R.id.bStatus).text  = booking.status.uppercase()
            item.findViewById<TextView>(R.id.bFare).text    = "P${booking.fare}"
            val btn = item.findViewById<Button>(R.id.bAcceptBtn)
            when (booking.status) {
                "pending" -> {
                    btn.visibility = View.VISIBLE; btn.text = "Accept"
                    btn.backgroundTintList = ColorStateList.valueOf(Color.parseColor("#10B981"))
                    btn.setOnClickListener { updateStatus(booking.id, "confirmed", btn) }
                }
                "confirmed" -> {
                    btn.visibility = View.VISIBLE; btn.text = "Pick up"
                    btn.backgroundTintList = ColorStateList.valueOf(Color.parseColor("#0EA5E9"))
                    btn.setOnClickListener { updateStatus(booking.id, "onboard", btn) }
                }
                "onboard" -> {
                    btn.visibility = View.VISIBLE; btn.text = "Drop off"
                    btn.backgroundTintList = ColorStateList.valueOf(Color.parseColor("#8B5CF6"))
                    btn.setOnClickListener { updateStatus(booking.id, "completed", btn) }
                }
                else -> btn.visibility = View.GONE
            }
            container.addView(item)
        }
    }

    private fun updateStatus(bookingId: Int, status: String, btn: Button) {
        btn.isEnabled = false; btn.text = "..."
        lifecycleScope.launch {
            try {
                val vId = if (status == "confirmed") vehicleId else null
                val res = ApiClient.service.updateBookingStatus(
                    bookingId, StatusUpdateRequest(status, vId))
                if (res.isSuccessful) runOnUiThread {
                    Toast.makeText(this@DriverActivity,
                        "Booking #$bookingId -> $status", Toast.LENGTH_SHORT).show()
                    fetchBookings()
                }
            } catch (e: Exception) {
                runOnUiThread { btn.isEnabled = true; fetchBookings() }
            }
        }
    }

    private fun startLocationUpdates() {
        if (ActivityCompat.checkSelfPermission(this,
                Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this,
                arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), 1001); return
        }
        val request = LocationRequest.Builder(Priority.PRIORITY_HIGH_ACCURACY, 3000).build()
        locationCallback = object : LocationCallback() {
            override fun onLocationResult(r: LocationResult) {
                r.lastLocation?.let {
                    SocketManager.sendLocationUpdate(vehicleId,
                        it.latitude, it.longitude, it.speed * 3.6f, it.bearing)
                    runOnUiThread {
                        findViewById<TextView>(R.id.speedValue)?.text =
                            (it.speed * 3.6).toInt().toString()
                    }
                }
            }
        }
        fusedLocationClient.requestLocationUpdates(request, locationCallback, Looper.getMainLooper())
    }

    private fun stopLocationUpdates() {
        if (::locationCallback.isInitialized)
            fusedLocationClient.removeLocationUpdates(locationCallback)
    }

    override fun onDestroy() {
        super.onDestroy()
        stopLocationUpdates()
        if (isOnline) SocketManager.goOffline(vehicleId)
    }
}
